src/category_theory/mittag_leffler.lean

```diff
@@ -0,0 +1,268 @@
+  simp only [map_id, types_id, set.range_id],
+  exact (set.range_iff_surjective.mpr $ h k j g).symm,
+end
+
+/--
+TODO: where does this go?
+-/
+lemma _root_.category_theory.is_cofiltered.cone_over_cospan
+  [is_cofiltered J] {i j j' : J} (f : j ⟶ i) (f' : j' ⟶ i)  :
+  ∃ (k : J) (g : k ⟶ j) (g' : k ⟶ j'), g ≫ f = g' ≫ f' :=
+begin
+  let h := is_cofiltered.min_to_left j j',
+  let h' := is_cofiltered.min_to_right j j',
+  let G := is_cofiltered.eq_hom (h ≫ f) (h' ≫ f'),
+  refine ⟨_, G ≫ h, G ≫ h', _⟩,
+  simp only [category.assoc, is_cofiltered.eq_condition],
+end
+
+lemma ranges_directed_of_is_cofiltered [is_cofiltered J] (j : J) :
+  directed_on (⊇) (set.range (λ ( f : Σ' (i : J), i ⟶ j), set.range (F.map f.2))) :=
+begin
+  rintros _ ⟨⟨i, ij⟩, rfl⟩ _ ⟨⟨k, kj⟩, rfl⟩,
+  obtain ⟨l, li, lk, e⟩ := category_theory.is_cofiltered.cone_over_cospan ij kj,
+  refine ⟨set.range (F.map $ li ≫ ij), _⟩,
+  rw [set.mem_range, exists_prop],
+  refine ⟨⟨⟨l, li ≫ ij⟩, rfl⟩, ⟨_, _⟩⟩,
+  rotate, rw e,
+  all_goals
+  { simp_rw [functor.map_comp, types_comp],
+    apply set.range_comp_subset_range, },
+end
+
+/--
+TODO: where does this go?
+-/
+private lemma directed_on_min {s : set J} [preorder J] (h : directed_on (≥) s)
+  (m ∈ s) (min : ∀ (a ∈ s), a ≤ m → a = m) : ∀ a ∈ s, m ≤ a :=
+λ a as, let ⟨x, xs, xm, xa⟩ := h m H a as in (min x xs xm) ▸ xa
+
+lemma is_mittag_leffler_of_exists_finite_range [is_cofiltered J]
+  (h : ∀ (j : J), ∃ i (f : i ⟶ j), (set.range (F.map f)).finite ) :
+  F.is_mittag_leffler :=
+begin
+  rintro j,
+  suffices : ∃ (f : Σ' i, i ⟶ j), ∀ (f' : Σ' i, i ⟶ j),
+               set.range (F.map f'.2) ≤ set.range (F.map f.2) →
+                 set.range (F.map f'.2) = set.range (F.map f.2),
+  { obtain ⟨⟨i, f⟩, fmin⟩ := this,
+    refine ⟨i, f, λ i' f', _⟩,
+    refine directed_on_min (F.ranges_directed_of_is_cofiltered j)
+                           _ ⟨⟨i, f⟩, rfl⟩ _ _ ⟨⟨i', f'⟩, rfl⟩,
+    simp only [set.mem_range, psigma.exists, forall_exists_index],
+    rintro _ k g rfl gf,
+    exact fmin ⟨k, g⟩ gf, },
+
+  let fins := subtype { f : Σ' i, i ⟶ j | (set.range (F.map f.2)).finite },
+  haveI : nonempty fins := by { obtain ⟨i, f, fin⟩ := h j, exact ⟨⟨⟨i, f⟩, fin⟩⟩, },
+  let fmin := function.argmin (λ (f : fins), f.prop.to_finset.card) nat.lt_wf,
+  use fmin.val,
+  rintro g gf,
+  cases lt_or_eq_of_le gf,
+  { have gfin : (set.range (F.map g.2)).finite := fmin.prop.subset gf,
+    refine ((λ (f : fins), f.prop.to_finset.card).not_lt_argmin nat.lt_wf ⟨g, gfin⟩ _).elim,
+    exact finset.card_lt_card (set.finite.to_finset_ssubset.mpr h_1), },
+  { assumption, },
+end
+
+/--
+The subfunctor of `F` obtained by restricting to the eventual range at each index.
+-/
+def to_eventual_ranges [is_cofiltered J] : J ⥤ Type v :=
+{ obj := λ j, F.eventual_range j,
+  map := λ i j f, set.maps_to.restrict (F.map f) _ _ ( by
+    { rintro x h,
+      simp only [eventual_range, set.mem_Inter, set.mem_range] at h ⊢,
```

Comment:
In this file, you use quite long `(by { ...})` blocks. `by` is recommended for very short proofs, otherwise `begin ... end` is more common in mathlib's coding style.